### ISSUE TITLE
Using "arguments" keyword with Function.prototype.apply causes TS2345

### ISSUE THREAD

**Issue by ghost** [microsoft/TypeScript#36470]
**TypeScript Version:**  3.7.5

**Search Terms:**
typescript function prototype apply arguments

**Code**

```ts
window.console.log = (() => {
    const originalFn = window.console.log;
    return function () {
        console.error('Called wrapped console.log()');
        return originalFn.apply(window.console, arguments);
    }
})();
console.log(1, 2, 3);
```

**Expected behavior:**
That snippet is valid JS code, it shouldn't cause an error.

**Actual behavior:**
It causes a TS2345 error because [IArguments](https://github.com/microsoft/TypeScript/blob/20e8eba143096cef7dff1529f7d850fc4c028494/lib/lib.es5.d.ts#L388) is not compatible with [T\[\]](https://github.com/microsoft/TypeScript/blob/20e8eba143096cef7dff1529f7d850fc4c028494/lib/lib.es5.d.ts#L335).

**Playground Link:** 
https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=7&pc=6#code/O4SwdgJg9sB0DGUwGcoBsCms1QOYAIBefAChIEoiA+fAbwFgAofF-RFAF3ygCcRdwAQzQAxMEXyhIMBElSZseANxNW+Hhg4BXHuIBmWsPA4gkpSg2ZrW7eVgw8evEgHIAwsMwRJPQQAc-DG9bdCwcXAoXchUra3VNHXFefiFRMFh-PzQATxIpaDgQzAAafEEeXC0AWwwwDmRo1VYAXyZm8golIA

**Related Issues:** 
\-

**Comment by RyanCavanaugh** [comment]
When `apply` synthesizes its argument type, it produces a tuple type like `[any?, ...args: any[]]`. Tuple types imply that they are actually implemented as `Array`s (e.g. you can use non-mutating methods like `slice`), but this isn't strictly necessary for `call`/`apply` - they can work on `ArrayLike`s, which make sense because `arguments` is only array-like (it's not really an array). So the more accurate type would be something like `ArrayLikeify<[any?, ...args: any[]]>` where `ArrayLikeify` would somehow just move over the non-array bits, but we're not super equipped to do this right now.

**Comment by isiahmeadows** [comment]
I have a local workaround for this:

```js
declare global {
    interface CallableFunction {
        apply<T, R>(
            this: (this: T, ...args: unknown[]) => R,
            thisArg: T, args: IArguments
        ): R;
    }
}
```

Obviously unsafe and I'm not sure if it's ideal for everyone, but it works well enough for me.

**Comment by zengyuanqiu** [comment]
### MDN
The apply() method calls the specified function with a given this value, and arguments provided as an array (or an [array-like object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#working_with_array-like_objects))

```javascript
Array.apply(null, {length: 3})
```

**Comment by zengyuanqiu** [comment]
### MDN
The apply() method calls the specified function with a given this value, and arguments provided as an array (or an [array-like object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#working_with_array-like_objects))

```javascript
Array.apply(null, {length: 3})
```
